driver-core/src/main/com/mongodb/internal/connection/OperationContext.java

```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.mongodb.internal.connection;
+
+import java.util.concurrent.atomic.AtomicLong;
+
+public class OperationContext {
+    private static final AtomicLong NEXT_ID = new AtomicLong(0);
+    private final long id;
+
+    public OperationContext() {
```

Comment:
It feels a bit unsettling just how many classes this gets threaded into, though I have no other suggestion.

Reply:
My first attempt at this was to make this a super-interface of SessionContext, but the refactoring required to do that got a bit out of control, so I backed off that.  But the motivation for that was exactly as you say, since SessionContext is already threaded into many of the places this is needed.

Reply:
For the SessionContext idea, the plan would have been to add `getOperationContext` to it, and store it in various implementers? And the issue was too many changes to implementers of SessionContext?

Reply:
I was going to make SessionContext extend OperationContext but there wouldn't have been any implementers of OperationContext only.  But the current split between ClusterBinding and ClientSessionBinding made that difficult.  I went down the road of combining those two classes but ran into some complications that I didn't feel were justified for this effort.